src/pip/_internal/utils/plugins.py

```diff
@@ -0,0 +1,72 @@
+import contextlib
+import logging
+from pathlib import Path
+from typing import Iterator, List
+
+from pip._vendor.pygments.plugin import iter_entry_points
```

Comment:
We shouldn't be using an undocumented helper function from pygments here. It could disappear without warning. If we need the functionality we should implement our own copy. Although isn't this available in `importlib.metadata` anyway?

Reply:
good point! I copied the logic to this file. The reason why we need it is because the [API for `EntryPoints`](https://docs.python.org/3/library/importlib.metadata.html#entry-points) changed in Python 3.10, so the way to get entry points matching a specific group name depends on the Python version:

```python
groups = entry_points()
if hasattr(groups, "select"):
    # New interface in Python 3.10 and newer versions of the
    # importlib_metadata backport.
    return groups.select(group=group_name)
else:
    assert hasattr(groups, "get")
    # Older interface, deprecated in Python 3.10 and recent
    # importlib_metadata, but we need it in Python 3.8 and 3.9.
    return groups.get(group_name, [])

```

Reply:
I'm so sick of the instability of the `importlib.{metadata, resources}` APIs, but we have to work with what we have, I guess :-(

Thanks for doing this.